### PR TITLE
UGENE-5263

### PR DESCRIPTION
**Pull request by rasputinkirill**
Circular mark does not cause auto-annotations update

### CODE REVIEW

**Review by DmitriiSukhomlinov — DISMISSED** [review]

**Review by yalgaer — CHANGES_REQUESTED** [review]
This is a very bad solution: you expect that ProjectViewImpl knows about every single usage of 'circular' flag in every view/widget.

A good solution is the following:
1) ProjectViewImpl updates 'circular' flag in U2SequenceObject.
2) U2SequenceObject emits 'circularChange' event (signal).
3) The code in AnnotatedDNAView responsible for auto-annotations listens for the signal and triggers annotations update if needed.

Please fix.

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 1]
```diff
@@ -785,6 +787,19 @@ void AnnotatedDNAView::sl_relatedObjectRelationChanged() {
     }
 }
 
+void AnnotatedDNAView::sl_sequenceCircularStateChanged() {
+    U2SequenceObject *o = qobject_cast<U2SequenceObject *>(sender());
+    CHECK(o != nullptr, );
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: nit: use safe-point

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 2]
```diff
@@ -785,6 +787,19 @@ void AnnotatedDNAView::sl_relatedObjectRelationChanged() {
     }
 }
 
+void AnnotatedDNAView::sl_sequenceCircularStateChanged() {
+    U2SequenceObject *o = qobject_cast<U2SequenceObject *>(sender());
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: nit: do not write object type twice in the same line: use auto

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 3]
```diff
@@ -785,6 +787,19 @@ void AnnotatedDNAView::sl_relatedObjectRelationChanged() {
...
+    CHECK(o != nullptr, );
+    for (ADVSequenceObjectContext *ctx : getSequenceContexts()) {
+        if (ctx->getSequenceObject() == o) {
+            QAction *toggleAAAction = AutoAnnotationUtils::findAutoAnnotationsToggleAction(ctx, ANNOTATION_GROUP_ENZYME);
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: nit: when a getter method says it is "Action" do not not use Action* but use auto

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_5001_6000.cpp [hunk 4]
```diff
@@ -1221,6 +1221,34 @@ GUI_TEST_CLASS_DEFINITION(test_5352) {
...
+    //1. Open _common_data/genbank/pBR322.gb
+    //2. Find the following restriction site: EcoRI
+    //Expected state: it is located on the sequence junction point
+    //3. Remove the circular mark
+    //Expected state: the restriction sites of sequence junction disappear
+    
+    GTFileDialog::openFile(os, dataDir + "samples/Genbank/PBR322.gb");
+    GTUtilsSequenceView::checkSequenceViewWindowIsActive(os);
+
+    GTUtilsDialog::waitForDialog(os, new PopupChooser(os, QStringList() << "ADV_MENU_ANALYSE"
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_5001_6000.cpp]: nit: use a shorter list construction: {A, B, C} (fix all places in the method)

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_5001_6000.cpp [hunk 5]
```diff
@@ -1221,6 +1221,34 @@ GUI_TEST_CLASS_DEFINITION(test_5352) {
...
+    CHECK_SET_ERR(region == "join(4359..4361,1..3)", QString("EcoRI region is incorrect: %1").arg(region));
+    CHECK_SET_ERR(GTUtilsAnnotationsTreeView::findItem(os, "EcoRI") != nullptr, QString("'EcoRI' item isn't found, but should."));
+
+    GTMouseDriver::moveTo(GTUtilsProjectTreeView::getItemCenter(os, "SYNPBR322"));
+    GTUtilsDialog::waitForDialog(os, new PopupChooserByText(os, QStringList() << "Mark as circular"));
+    GTMouseDriver::click(Qt::RightButton);
+    GTUtilsTaskTreeView::waitTaskFinished(os);
+    GTGlobals::FindOptions options;
+    options.failIfNotFound = false;
+    CHECK_SET_ERR(GTUtilsAnnotationsTreeView::findItem(os, "EcoRI", options) == nullptr, QString("'EcoRI' item is found, but should not."));
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_5001_6000.cpp:1249]: nit: add a backward check too

File: src/ugeneui/src/project_view/ProjectViewImpl.cpp [hunk 6]
```diff
@@ -1175,6 +1176,7 @@ void ProjectViewImpl::sl_onToggleCircular() {
         const bool objectIsModifiable = (!obj->isStateLocked() && !projectTreeController->isObjectInRecycleBin(obj));
         if (objectIsModifiable && obj->getGObjectType() == GObjectTypes::SEQUENCE) {
             U2SequenceObject *casted = qobject_cast<U2SequenceObject *>(obj);
+            CHECK(casted != nullptr);
```
Comment (yalgaer) [inline src/ugeneui/src/project_view/ProjectViewImpl.cpp]: nit: use SAFE_POINT

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 7]
```diff
@@ -908,6 +923,7 @@ QString AnnotatedDNAView::addObject(GObject *o) {
         addRelatedAnnotations(sc);
         emit si_sequenceAdded(sc);
         connect(o, SIGNAL(si_relatedObjectRelationChanged()), SLOT(sl_relatedObjectRelationChanged()));
+        connect(o, SIGNAL(si_sequenceCircularStateChanged()), SLOT(sl_sequenceCircularStateChanged()));
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: where is the disconnect?

**Review by yalgaer — COMMENTED** [review]
Minor comments only: please check.
Will approve after all tests pass

**Review by yalgaer — COMMENTED** [review]
Looks good. Will approve after all tests are finished.

**Comment by yalgaer** [comment]
Ping. I will approve only after GUI/XML  run

**Comment by yalgaer** [comment]
Closing this pull request as "inactive". Please re-open when you are ready.

**Comment by rasputinkirill** [comment]
Al tests passed.

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp [hunk 8]
```diff
@@ -4994,61 +4994,6 @@ GUI_TEST_CLASS_DEFINITION(test_3809) {
     GTUtilsTaskTreeView::waitTaskFinished(os);
 }
 
-GUI_TEST_CLASS_DEFINITION(test_3813) {
```
Comment (yalgaer) [inline src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp]: Why this test is removed? Which code in UGENE does not work after changes in this patch?

**Review by yalgaer — COMMENTED** [review]
Please add a description why the test is removed

File: src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp (hunk 8 above)
Comment (rasputinkirill) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp]: Unload option for documents removed from ugene. 
ProjectTreeController.cpp
image

Comment (rasputinkirill) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp]: This test also failed on my machine on latest master with corresponding error ` GT_FAIL: (action != nullptr) for 'GTMenu __ clickMenuItem _ action not found for item action_project__unload_selected_action'`

Comment (yalgaer) [reply src/plugins/GUITestBase/src/tests/regression_scenarios/GTTestsRegressionScenarios_3001_4000.cpp]: This action is present in UGENE only in GUI test mode.
Run it like 'ugeneui --gui-test=test-name' from the command line.

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 9]
```diff
@@ -785,6 +788,19 @@ void AnnotatedDNAView::sl_relatedObjectRelationChanged() {
     }
 }
 
+void AnnotatedDNAView::sl_sequenceCircularStateChanged() {
+    auto *o = qobject_cast<U2SequenceObject *>(sender());
+    SAFE_POINT(o != nullptr, "casting to 'U2SequenceObject' failed", );
+    for (ADVSequenceObjectContext *ctx : getSequenceContexts()) {
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: use qAsConst with for-loops and Qt collections

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 10]
```diff
@@ -785,6 +788,19 @@ void AnnotatedDNAView::sl_relatedObjectRelationChanged() {
     }
 }
 
+void AnnotatedDNAView::sl_sequenceCircularStateChanged() {
+    auto *o = qobject_cast<U2SequenceObject *>(sender());
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: use meaningful names: rename 'o' to  'sequenceObject'

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp [hunk 11]
```diff
@@ -908,6 +924,7 @@ QString AnnotatedDNAView::addObject(GObject *o) {
         addRelatedAnnotations(sc);
         emit si_sequenceAdded(sc);
         connect(o, SIGNAL(si_relatedObjectRelationChanged()), SLOT(sl_relatedObjectRelationChanged()));
+        connect(o, SIGNAL(si_sequenceCircularStateChanged()), SLOT(sl_sequenceCircularStateChanged()));
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.cpp]: Use function pointer based signal slot connectors for all new connect calls

File: src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.h [hunk 12]
```diff
@@ -228,6 +228,7 @@ private slots:
     void sl_aminoTranslationChanged();
     void sl_updatePasteAction();
     void sl_relatedObjectRelationChanged();
+    void sl_sequenceCircularStateChanged();
```
Comment (yalgaer) [inline src/corelibs/U2View/src/ov_sequence/AnnotatedDNAView.h:232]: Add docs for all new methods

**Review by yalgaer — CHANGES_REQUESTED** [review]
Consistent style issues only (except the removed test)

**Comment by rasputinkirill** [comment]
Tests passed.

**Review by yalgaer — APPROVED** [review]

**Pull request merged**